README.md
# node-easel

node-easel is a node wrapper for [EaselJS](https://github.com/CreateJS/EaselJS).
For use with [NodeJS](http://nodejs.org), built on-top of [node-canvas](https://github.com/LearnBoost/node-canvas)

## Installation

	$ npm install node-easel

**Note**
Cairo graphics is required to run node-easel && node-canvas.
Read the install docs at https://github.com/LearnBoost/node-canvas, for full install instructions.

## Examples

To see a full working demo, checkout the [examples](examples/) folder.


## Simple Example

node-easel is completely polymorphic with EaselJS. A good starting point is to checkout the [EaselJS documentation](http://createjs.com/Docs/EaselJS/).

```javascript
//Import easel
require('node-easel');
var Stage = createjs.Stage;
var Shape = createjs.Shape;
var Graphics = createjs.Graphics;

var fs = require('fs');

//Create the canvas to draw to
var c = new Canvas(980, 580);
var ctx = c.getContext('2d');

//Create graphics object
var g = new createjs.Graphics();
var shape = new createjs.Shape(g);

//Draw a circle
g.setStrokeStyle(8)
.beginStroke("#F0F")
.beginRadialGradientFill(["#FF0","#00F"],[0,1],100,200,0,100,200,40)
.drawCircle(100,200,40);

//Add the item to our stage, and call .tick(); to draw the object.
var stage = new createjs.Stage(c);
stage.addChild(shape);
stage.tick();

//Create a PNG file.
fs.writeFile(__dirname + '/public/circle.png', c.toBuffer(), function() {
	createjs.Ticker.halt();
});
```


HISTORY.md
0.1.2 / 2013-11-21
==================

	* Parity update for EaselJS 0.7.0
	* Due to API changes in EaselJS this version will break old projects.
	* * For details checkout the EaselJS changelog; https://github.com/CreateJS/EaselJS/blob/master/VERSIONS.txt

0.1.1 / 2013-05-15
==================

	* Update for parity with EaselJS 0.6.1.


0.1.0 / 2013-02-14
==================

	* Initial Release. Parity with EaselJS 0.6.0.


src/node-easel.js
/*
 * Copyright (c) 2013 Wes Gorgichuk
 *
 * Permission is hereby granted, free of charge, to any person
 * obtaining a copy of this software and associated documentation
 * files (the "Software"), to deal in the Software without
 * restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following
 * conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 */
var Canvas = require('canvas');
var Image = Canvas.Image;

/**
 * Surpress addEventListener errors on easel.
 * Its currently only used for MouseEvent, so its not needed on the server.
 *
 */
Canvas.prototype.addEventListener = function () { };

/**
 * Inject a window object
 *
 * @type {Object}
 */
window = { addEventListener:function () { } };

/**
 * node-canvas doesn't support cloneNode();
 * So create our own.
 *
 * @return {Canvas}
 */
Canvas.prototype.cloneNode = function () {
	var c = new Canvas(this.width, this.height);
	c.type = this.type;

	return c;
};

// Easel uses instanceof HTMLCanvasElement, so change it to Canvas.
HTMLCanvasElement = Canvas;

// Create our global createjs namespace.
createjs = {
	_snapToPixelEnabled:true,

	createCanvas:function () {
		return new Canvas();
	},

	createImage:function () {
		return new Image();
	}
};

var classes = [
	// Shared
	'createjs/events/EventDispatcher',
	'createjs/events/Event',
	'createjs/utils/IndexOf',

	// TweenJS code (used by MovieClip)
	'tweenjs/CSSPlugin',
	'tweenjs/Ease',
	'tweenjs/MotionGuidePlugin',
	'tweenjs/Timeline',
	'tweenjs/Tween',
	'tweenjs/version',

	// EaselJS code
	'easeljs/utils/UID',
	'easeljs/utils/SpriteSheetBuilder',
	'easeljs/utils/SpriteSheetUtils',
	'easeljs/utils/Ticker',
	'easeljs/events/MouseEvent',
	'easeljs/geom/Matrix2D',
	'easeljs/geom/Rectangle',
	'easeljs/geom/Point',
	'easeljs/display/DisplayObject',
	'easeljs/display/Container',
	'easeljs/display/Stage',
	'easeljs/display/Shadow',
	'easeljs/display/Shape',
	'easeljs/display/SpriteSheet',
	'easeljs/display/Sprite',
	'easeljs/display/Text',
	'easeljs/display/Bitmap',
	'easeljs/display/BitmapText',
	'easeljs/display/BitmapAnimation',
	'easeljs/display/Graphics',
	'easeljs/display/MovieClip',
	'easeljs/filters/Filter',
	'easeljs/filters/AlphaMapFilter',
	'easeljs/filters/AlphaMaskFilter',
	'easeljs/filters/BlurFilter',
	'easeljs/filters/ColorFilter',
	'easeljs/filters/ColorMatrix',
	'easeljs/filters/ColorMatrixFilter',
	'easeljs/version',
	'easeljs/version_movieclip'
];

for (var i = 0; i < classes.length; i++) {
	var path = classes[i];
	var name = path.split('/').pop();
	require('./' + path + '.js')[name];
};

/**
 * Inject custom functionality that is only required on the server.
 * So we can keep the same EaselJS source desktop / server.
 *
 */

/**
 * Inject a halt method for Ticker.
 * Clears the Ticker's Timeout, and stops all animation.
 * Should only be called when your ready to stop the node instance.
 *
 */
createjs.Ticker.halt = function() {
	if (createjs.Ticker.timeoutID !== null) {
		clearTimeout(createjs.Ticker.timeoutID);
	}
}


examples/server.js
var express = require('express')
var app = express();
var fs = require('fs');
var exec = require('child_process').exec;

var captcha = require('./captcha.js').Captcha;
var graphicsTest = require('./GraphicsTest.js').GraphicsTest;
var spriteSheetBuilderTest = require('./SpriteSheetBuilderTest').SpriteSheetBuilderTest;
var spritesheetTest = require('./SpritesheetTest').SpritesheetTest;

//app.use(express.favicon());
app.use(express.logger('dev'));
app.use(express.bodyParser());
app.use(app.router);
app.use(express.static(__dirname + '/public'));
app.set('views', __dirname + '/public');
app.use(express.errorHandler({ showStack:true }));

app.get('/captcha.png', function (req, res) {
	var seed = parseInt(req.query.seed || Math.random() * 0xffffff);

	var data = new captcha(seed);
	sendImage(data.buffer, res);
});

app.get('/graphicsTest.png', function (req, res) {
	var data = new graphicsTest();
	sendImage(data.buffer, res);
});

app.get('/spritesheetBuilder.png', function (req, res) {
	var data = new spriteSheetBuilderTest();
	sendImage(data.buffer, res);
});

app.get('/running.mpg', function (req, res) {
	exec('which ffmpeg', function (err, result) {
		if (result.length != 0) {
			new spritesheetTest(function (videoPath) {
				res.contentType('video/mpeg');
				res.sendfile(__dirname + '/output/' + videoPath);
			});
		} else {
			res.set('Content-Type', 'text/html');
			res.send(new Buffer('Error: ffmpeg is required for this demo.  Install from <a href="http://www.ffmpeg.org/">ffmpeg.org</a> '));
		}
	});


});

function sendImage(buffer, res) {
	res.contentType('image/png');

	buffer.toBuffer(function (err, buf) {
		res.send(buf);
	});
}


app.get('/', function (req, res) {
	res.sendfile(__dirname + '/public/index.html');
});

app.listen(9000);

var os = require('os')

var interfaces = os.networkInterfaces();
var addresses = [];
for (var n in interfaces) {
	for (var j in interfaces[n]) {
		var address = interfaces[n][j];
		if (address.family == 'IPv4' && !address.internal) {
			addresses.push(address.address)
		}
	}
}

console.log('App started: ', addresses)


examples/captcha.js
/*
 * Copyright (c) 2013 Wes Gorgichuk
 *
 * Permission is hereby granted, free of charge, to any person
 * obtaining a copy of this software and associated documentation
 * files (the "Software"), to deal in the Software without
 * restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following
 * conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 */
var fs = require('fs');
var Canvas = require('canvas');
var Image = Canvas.Image;
var Rnd = require('./public/com/gskinner/utils/Rnd.js').Rnd;

require('../src/node-easel.js');

var Captcha = function (seed) {
	return this.init(seed);
}
exports.Captcha = Captcha;

var p = Captcha.prototype = {};

p.init = function (seed) {
	var canvas = new Canvas();
	var ctx = canvas.getContext('2d');

	Rnd.setSeed(seed);

	var c = new Canvas(175, 75);
	var ctx = c.getContext('2d');
	var stage = new createjs.Stage(c);

	var usedCharacters = [];

	var characters = 'abcdefjhjklmnpqrxtuvwxyz23456789'.split('');

	var length = 9;
	var lastX = 15;

	while (length--) {
		var char = characters[Rnd.random() * characters.length | 0];
		usedCharacters.push(char);

		var t = new createjs.Text(char, (25) + "px Verdana", "#000");
		t.textBaseline = "top";

		var w = 20;//ctx.measureText(char).width;
		var h = 25;//t.getMeasuredLineHeight();

		t.x = lastX;
		t.y = Rnd.random() * (c.height - h);
		t.skewX = Rnd.random() * w * 2;
		t.skewY = Rnd.random() * h * 2;

		stage.addChild(t);

		lastX += w + (Rnd.random() * w);
	}

	var g = new createjs.Graphics();
	g.initialize(ctx);

	var shape = new createjs.Shape(g);
	g.setStrokeStyle(1).beginStroke("#00acc");

	length = 10;
	while (length--) {
		g.lineTo(Rnd.random() * c.width, Rnd.random() * c.height);
		g.moveTo(Rnd.random() * c.width, Rnd.random() * c.height);
	}

	var g2 = new createjs.Graphics();
	g2.initialize(ctx);

	var shape2 = new createjs.Shape(g2);
	length = 40;
	while (length--) {
		g2.setStrokeStyle(0).
				beginFill('#' + (Rnd.random() * 0xffffff | 0)).
				drawCircle(Rnd.random() * c.width, Rnd.random() * c.height,
						   Rnd.random() * 3 | 0)
				.endFill();
	}

	stage.addChild(shape, shape2);

	var blurFilter = new createjs.BlurFilter(3, 2, 2);
	var margins = blurFilter.getBounds();
	t.cache(0,
			0,
			t.getMeasuredWidth() + margins.width,
			t.getMeasuredLineHeight() + margins.height);

	stage.update();

	return {buffer:c,
		term:usedCharacters.join('')};
}


examples/GraphicsTest.js
var fs = require('fs');
var Canvas = require('canvas');
var Image = Canvas.Image;

require('../src/node-easel.js');

var GraphicsTest = function () {
	return this.init();
}
var p = GraphicsTest.prototype = {};

exports.GraphicsTest = GraphicsTest;

p.init = function () {
	this.canvas = new Canvas(1042, 400);

	// create a new stage and point it at our canvas:
	this.stage = new createjs.Stage(this.canvas);

	// grab canvas width and height for later calculations:
	this.w = this.canvas.width;
	this.h = this.canvas.height;

	this.img = new Image();
	this.img.src = fs.readFileSync(__dirname + '/img/daisy.png');

	this.layout();

	return {buffer:this.canvas, term:'Easel Demo'};
}

p.layout = function () {
	var arr = [
		this.createStar,
		this.createHex,
		this.createLineTo,
		this.createRadialGradientFill,
		this.createEllipse,
		this.createRectGradientFill,
		this.createBitmapFill,
		this.createGreyImage,
		this.createColorImage,
		this.createBlurImage,
		this.createSmiley,
		this.createText
	];
	var padding = 5;
	var _width = 155;
	var _height = 155;
	var cols = 6;
	var space = 0;
	var l = arr.length;

	var bgContainer = new createjs.Container();
	var bg = new createjs.Shape();
	bg.graphics.beginFill('#333').drawRect(0, 0, this.w, this.h).endFill();
	bgContainer.addChild(bg);
	this.stage.addChild(bgContainer);

	for (var i = 0; i < l; i++) {
		var func = arr[i].bind(this);
		var tile = func();
		tile.x = 42 + (_width + padding) * (i % cols);
		tile.y = 42 + (i / cols | 0) * (_height + padding);
		this.stage.addChild(tile);
	}

	var border = this.createBorder();
	this.stage.addChild(border);
	this.stage.update();
}

p.createSmiley = function () {
	var s = new createjs.Shape();
	var g = s.graphics;

	//Head
	g.setStrokeStyle(15, 'round', 'round');
	g.beginStroke("#000");
	g.beginFill("#F00");
	g.drawCircle(170, 170, 170); //55,53
	g.endFill();
	g.setStrokeStyle(1, 'round', 'round');

	//Right eye
	g.setStrokeStyle(5, 'round', 'round');
	g.beginStroke("#000");
	g.beginFill("#000");
	g.drawRoundRect(125, 64, 20, 50, 10);
	g.endFill();

	//Left eye
	g.setStrokeStyle(5, 'round', 'round');
	g.beginStroke("#000");
	g.beginFill("#000");
	g.drawRoundRect(200, 64, 20, 50, 10);
	g.endFill();

	//Mouth
	g.setStrokeStyle(15, 'round', 'round');
	g.beginStroke("#000");
	g.moveTo(45, 155);
	g.bezierCurveTo(83, 307, 254, 317, 296, 152);

	var c = new createjs.Container();
	c.addChild(s);

	c.scaleX = .4;
	c.scaleY = .4;

	c.x = 10;
	c.y = 10;

	var container = this.createTile();
	container.addChild(c);

	return container;
}

p.createText = function () {
	var container = this.createTile();
	var text = new createjs.Text("Hello World",
								 "20px bold Arial",
								 "#00ffaa");
	text.x = 10;
	text.y = 50;
	container.addChild(text);
	container.shadow = new createjs.Shadow(createjs.Graphics.getRGB(255, 0, 0, 1), 5, 5, 5);
	return container;
}

p.createGreyImage = function () {
	var greyScaleFilter = new createjs.ColorMatrixFilter([
															 0.33, 0.33, 0.33, 0, 0, // red
															 0.33, 0.33, 0.33, 0, 0, // green
															 0.33, 0.33, 0.33, 0, 0, // blue
															 0, 0, 0, 1, 0  // alpha
														 ]);
	var bitmap = this.createBitmap(this.img, [greyScaleFilter]);
	var container = this.createTile();
	container.addChild(bitmap);
	return container;
}

p.createColorImage = function () {
	var bitmap = this.createBitmap(this.img, [new createjs.ColorFilter(0, 1, 1, 1)]);
	var container = this.createTile();
	container.addChild(bitmap);
	return container;
}

p.createBlurImage = function () {
	var blurFilter = new createjs.BlurFilter(5, 2, 2);
	var bitmap = this.createBitmap(this.img, [blurFilter]);
	var container = this.createTile();
	container.addChild(bitmap);
	return container;
}

p.createBitmap = function (image, filters) {
	var bitmap = new createjs.Bitmap(image);
	bitmap.filters = filters;
	bitmap.cache(0, 0, image.width, image.height);
	bitmap.x = 40;
	bitmap.y = 40;
	return bitmap;
}

p.createBorder = function () {
	var container = new createjs.Container();
	var s = new createjs.Shape();
	s.graphics.beginBitmapStroke(this.img)
			.setStrokeStyle(32)
			.drawRect(20, 20, 1000, 360);
	container.addChild(s);
	return container;
}

p.createBitmapFill = function () {
	var container = this.createTile();
	var s = new createjs.Shape();
	s.graphics.beginBitmapFill(this.img)
			.setStrokeStyle(8)
			.beginRadialGradientStroke(["#FFF", "#000"],
									   [0, 1],
									   0, 0, 0, 0, 30, 130)
			.drawRect(0, 0, 130, 130);
	s.x = 12;
	s.y = 10;
	container.addChild(s);
	return container;
}

p.createRectGradientFill = function () {
	var container = this.createTile();
	var s = new createjs.Shape();
	s.graphics.beginLinearGradientFill(["#FFF", "#000"],
									   [0, 1], 0, 0, 0, 130)
			.drawRect(0, 0, 130, 130);
	s.x = 12;
	s.y = 10;
	container.addChild(s);
	return container;
}

p.createEllipse = function () {
	var container = this.createTile();
	var s = new createjs.Shape();
	s.graphics.f(createjs.Graphics.getRGB(0, 0x66, 0x99, 0.5))
			.setStrokeStyle(4)
			.beginLinearGradientStroke(["#F00", "#000"],
									   [0, 1],
									   0, 0, 70, 140)
			.drawEllipse(0, 0, 70, 140, 8);
	s.x = 40;
	s.y = 10;
	container.addChild(s);
	return container;
}

p.createRadialGradientFill = function () {
	var container = this.createTile();
	var s = new createjs.Shape();
	s.graphics.ss(8)
			.beginStroke("#f0f")
			.beginRadialGradientFill(["#FFF", "#0FF"],
									 [0, 1],
									 0, 0, 0, 0, 0, 40)
			.drawCircle(0, 0, 40);
	s.x = s.y = 80;
	container.addChild(s);
	return container;
}

p.createLineTo = function () {
	var container = this.createTile();
	var s = new createjs.Shape();
	s.graphics.setStrokeStyle(16, "round", "round")
			.beginStroke("#f90")
			.moveTo(20, 10)
			.lineTo(90, 90)
			.lineTo(90, 140);
	container.addChild(s);
	return container;
}

p.createHex = function () {
	var container = this.createTile();
	var s = new createjs.Shape();
	s.graphics.beginFill("#0F0")
			.drawPolyStar(0, 0, 40, 6)
			.drawPolyStar(0, 75, 40, 6);
	s.x = 80
	s.y = 40;

	container.addChild(s);
	return container;
}

p.createStar = function () {
	var container = this.createTile();
	var s = new createjs.Shape();
	s.graphics.setStrokeStyle(1)
			.beginStroke(createjs.Graphics.getRGB(255, 255, 0))
			.beginFill("#FF0")
			.endStroke()
			.drawPolyStar(0, 0, 80, 5, 0.6, -90);
	s.x = 80
	s.y = 85;

	container.addChild(s);
	return container;
}

p.createTile = function () {
	var container = new createjs.Container();
	var bg = new createjs.Shape();
	bg.graphics.beginFill('#CCCCCC').drawRect(0, 0, 155, 155).endFill();
	bg.alpha = 0.25;
	container.addChild(bg);
	return container;
}


examples/public/com/gskinner/utils/Rnd.js
/*
 * Copyright (c) 2010 Grant Skinner
 *
 * Permission is hereby granted, free of charge, to any person
 * obtaining a copy of this software and associated documentation
 * files (the "Software"), to deal in the Software without
 * restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following
 * conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 */
(function (window) {

	var Rnd = function () {
		throw new Error('Rnd is static and cannot be instantiated.');
	};

	Rnd.setSeed = function (value) {
		Rnd._currentSeed = value;
	};

	Rnd.randFloat = function (min, max) {
		if (isNaN(max)) {
			max = min;
			min = 0;
		}
		return Math.random() * (max - min) + min;
	};

	// boolean(); // returns true or false (50% chance of true)
	// boolean(0.8); // returns true or false (80% chance of true)
	Rnd.randBoolean = function (chance) {
		if (isNaN(chance)) {
			chance = .5;
		}
		return (Math.random() < chance);
	};

	// sign(); // returns 1 or -1 (50% chance of 1)
	// sign(0.8); // returns 1 or -1 (80% chance of 1)
	Rnd.randSign = function (chance) {
		if (isNaN(chance)) {
			chance = .5;
		}
		return (Math.random() < chance) ? 1 : -1;
	}

	// bit(); // returns 1 or 0 (50% chance of 1)
	// bit(0.8); // returns 1 or 0 (80% chance of 1)
	Rnd.randBit = function bit(chance) {
		if (isNaN(chance)) {
			chance = .5;
		}
		return (Math.random() < chance) ? 1 : 0;
	};

	// integer(50); // returns an integer between 0-49 inclusive
	// integer(20,50); // returns an integer between 20-49 inclusive
	Rnd.randInteger = function (min, max) {
		if (isNaN(max)) {
			max = min;
			min = 0;
		}
		// Need to use floor instead of bit shift to work properly with negative values:
		return Math.floor(Rnd.randFloat(min, max));
	};

	Rnd.random = function () {
		var rand = (Rnd._currentSeed = (Rnd._currentSeed * 16807) % 2147483647) / 0x7FFFFFFF + 0.000000000233;
		return rand;
	}

	window.Rnd = Rnd;

}((typeof exports == 'undefined') ? window : exports));


examples/SpritesheetTest.js
var Canvas = require('canvas');
var Image = Canvas.Image;
var fs = require('fs');
var exec = require('child_process').exec;

require('../src/node-easel.js');

var SpritesheetTest = function (success) {
	this.init(success);
}
exports.SpritesheetTest = SpritesheetTest;

var p = SpritesheetTest.prototype = {};
var s = SpritesheetTest;

s.RUNNING_RATE = 2.5;

p.init = function (success) {
	var cmd = "find " + __dirname + "/output -name '*.mpg' -exec rm -f {} \\;";
	console.log(cmd);
	exec(cmd);

	this.success = success;

	this.canvas = new Canvas(960, 400);
	this.stage = new createjs.Stage(this.canvas);

	this.index = 0;

	var spriteSheet = {"animations":{"run":[0, 25], "jump":[26, 63]}, "images":[this.loadImage("runningGrant.png")], "frames":{"regX":0, "height":292.5, "count":64, "regY":0, "width":165.75}};

	var ss = new createjs.SpriteSheet(spriteSheet);
	this.grant = new createjs.BitmapAnimation(ss);

	// Set up looping
	ss.getAnimation("run").next = "run";
	ss.getAnimation("jump").next = "run";
	this.grant.gotoAndPlay("run");

	// Position the Grant sprite
	this.grant.x = -200;
	this.grant.y = 90;
	this.grant.scaleX = this.grant.scaleY = 0.8;

	// grab canvas width and height for later calculations:
	this.w = this.canvas.width;
	this.h = this.canvas.height;

	var sky = new createjs.Shape(new createjs.Graphics().beginBitmapFill(this.loadImage('sky.png')).drawRect(0, 0, this.w, this.h));

	this.ground = new createjs.Shape();
	var g = this.ground.graphics;
	g.beginBitmapFill(this.loadImage('ground.png'));
	g.drawRect(0, 0, this.w + 330, 79);
	this.ground.y = this.h - 79;

	this.hill = new createjs.Shape(new createjs.Graphics().beginBitmapFill(this.loadImage('parallaxHill1.png')).drawRect(0, 0, 282, 59));
	this.hill.x = Math.random() * this.w;
	this.hill.scaleX = this.hill.scaleY = 3;
	this.hill.y = 144;

	this.hill2 = new createjs.Shape(new createjs.Graphics().beginBitmapFill(this.loadImage('parallaxHill2.png')).drawRect(0, 0, 212, 50));
	this.hill2.x = Math.random() * this.w;
	this.hill2.scaleX = this.hill2.scaleY = 3;
	this.hill2.y = 171;

	this.stage.addChild(sky, this.ground, this.hill, this.hill2, this.grant);

	this.tickFunction = this.tick.bind(this);
	var fps = 80;
	createjs.Ticker.setFPS(fps);
	createjs.Ticker.addEventListener("tick", this.tickFunction);

	var seconds = 10;
	var millseconds = seconds * 1000;
	console.log(millseconds);
	setTimeout(this.stopCapture.bind(this), millseconds);
}

p.stopCapture = function () {
	createjs.Ticker.removeEventListener("tick", this.tickFunction);
	console.log('begin video encoding');
	exec('cd ' + __dirname + '/output && ffmpeg -f image2 -i test_%d.png -sameq video.mpg', this.handleVideoEncoded.bind(this));
}

p.handleVideoEncoded = function () {
	console.log('Video Encoded');

	var cmd = "find " + __dirname + "/output -name '*.png' -exec rm -f {} \\;";
	exec(cmd);
	this.success('video.mpg');

	createjs.Ticker.halt();
}

p.loadImage = function (name) {
	var img = new Image();
	img.src = fs.readFileSync(__dirname + '/img/' + name);
	return img;
}

p.handleJumpStart = function () {
	this.grant.gotoAndPlay("jump");
}

p.tick = function () {
	var outside = this.w + 20;
	var position = this.grant.x + s.RUNNING_RATE;
	this.grant.x = (position >= outside) ? -200 : position;

	this.ground.x = (this.ground.x - 15) % 330;
	this.hill.x = (this.hill.x - 0.8);
	if (this.hill.x + 838 <= 0) {
		this.hill.x = outside;
	}
	this.hill2.x = (this.hill2.x - 1.2);
	if (this.hill2.x + 633 <= 0) {
		this.hill2.x = outside;
	}

	this.stage.update();

	var fileName = __dirname + '/output/test_' + (this.index++) + '.png';
	fs.writeFileSync(fileName, this.canvas.toBuffer(), '');
}
